ether-wake.c

```diff
@@ -278,6 +291,11 @@ static int send_magic_packet()
 	}
 #endif
 
+	// Wait until host is reachable again
+	if (hold) {
+		hold_for_online();
+	}
```

Comment:
I think we shouldn't couple the method send_magic_packet() to the higher level concept of hold; I would compose the control flow in the program's main loop - or implement a new function to be used as the callback which does the online check (if option hold is used) and then conditionally call `send_magic_packet()`.

Also, I think that the ping should return right away not wait for too long, so that the main loop can continue consuming data from the socket.

Btw, for purposes of holding back magic packets (to not unnecessarily send a lot of them), I have some suggestions in the README of the corresponding [OpenWrt package repository](https://github.com/mister-benjamin/etherwake-nfqueue-openwrt#setup-firewall-rules):

> The [iptables] option --match conntrack --ctstate NEW only matches packets of a new connection and --match limit --limit 3/hour --limit-burst 1 limits the amount of packets that are matched. The latter option roughly matches only one packet per 20 minutes. The intention here is to not be too intrusive and avoid sending a lot of magic packets.

Reply:
> I can implement another callback for hold, which also includes calling the magic packet function.

Yeah, that would be nice. Depending on configuration, one or the other could be registered.

> Well, I mean, what would the process do with the consumed packets if it does not know if the host is up or not. Since one instance of etherwake-nfqueue can only handle one destination, the program needs to know if the host is online or not. I'm not sure if asynchronicity would benefit the program?

There’s not necessarily a requirement for concurrency, but as the callback is called from the read loop of the nfqueue socket, I believe the callback shouldn’t block for too long as incoming data on the socket would pile up without being consumed. An easy way around this would be my suggestion in the other comment to set up and tear down the nfqueue client depending on the host’s online state. Otherwise maybe an atomic bool could be updated from a ping thread or only one icmp packet per callback could be sent.

Reply:
PS: I think not consuming messages from the nfqueue socket also means that the filtered network packets are stuck in the netfilter subsystem until the verdict has been issued, which would also be a strong argument to not block the read loop (for too long). Maybe it would even be better to not send the magic packet in the callback itself, but to only set a flag that it should be send.

Reply:
Sure, concurrency makes sense, when the packet is only considered an event for etherwake-nfqueue to act on. This is currently the case and there I can see the approach of the packet only activating a flag to send a magic packet.

However, in my use case I would like to hold back packets until I've woken up the host and received a ping. If we were to release the packet before, the host will not receive it and the client has to send a new packet. With my approach, concurrency makes less sense, since all packets in the queue require the host to be up. Thus, it is preferable to hold them in the queue then sending them to an unresponsive host.

Reply:
I think I am beginning to understand now: your whole point is to buffer up any TCP or UDP packets sent to an offline host and release them by sending the nfqueue verdict only after the host is online?

Reply:
Out of curiosity, what’s exactly your use case? Isn’t that maybe an abuse of the netfilter/nfqueue subsystem with potential problems when it’s taking some time until the host is up and a lot of traffic is coming in? It probably also means that you can’t use any of the iptables rule optimisations like “new connections only” or “rate limiting” and every TCP or UDP packet will pass through etherwake-nfqueue which will send a magic packet and issue a verdict?